pmlc/rt/vulkan/vulkan_invocation.cc

```diff
@@ -789,29 +789,32 @@ void VulkanInvocation::submitCommandBuffersToQueue() {
                      "vkQueueSubmit");
 }
 
-void VulkanInvocation::updateHostMemoryBuffers() {
-  for (const auto &action : schedule) {
-    if (auto kernel = std::dynamic_pointer_cast<LaunchKernelAction>(action)) {
-      // For each descriptor set.
-      for (auto &resourceDataMapPair : kernel->resourceData) {
-        auto &resourceDataMap = resourceDataMapPair.second;
-        auto &deviceMemoryBuffers =
-            kernel->deviceMemoryBufferMap[resourceDataMapPair.first];
-        // For each device memory buffer in the set.
-        for (auto &deviceMemoryBuffer : deviceMemoryBuffers) {
-          if (resourceDataMap.count(deviceMemoryBuffer.bindingIndex)) {
-            void *payload;
-            auto &hostMemoryBuffer =
-                resourceDataMap[deviceMemoryBuffer.bindingIndex];
-            throwOnVulkanError(vkMapMemory(device->getDevice(),
-                                           deviceMemoryBuffer.deviceMemory, 0,
-                                           hostMemoryBuffer.size, 0,
-                                           reinterpret_cast<void **>(&payload)),
-                               "vkMapMemory");
-            std::memcpy(hostMemoryBuffer.ptr, payload, hostMemoryBuffer.size);
-            vkUnmapMemory(device->getDevice(), deviceMemoryBuffer.deviceMemory);
-          }
-        }
+void VulkanInvocation::updateLastShaderHostMemoryBuffers() {
+  const auto &action = schedule.back();
+  auto kernel = std::dynamic_pointer_cast<LaunchKernelAction>(action);
+  if (!kernel) {
```

Comment:
Is there a different way to handle this case?  Throwing here seems pretty severe.  Could we search from back-toward-front until we find a LaunchKernelAction?

Reply:
Yes, I agree we can search back to front and avoid throws exception. This patch is sort of a workaround I think we should check and map all buffers needed not only the last a few buffers as Jeremy mentioned in the meeting.

Reply:
Yes.  We do know which buffers are user buffers (passed in as arguments) vs locals (alloced within the original function), so we should base the decision on that information rather than kernel #.  I'd rather not commit this change as as.

Reply:
Thanks, right now map only user buffers. Resnet50 inference latency improved from about 650ms to 300ms. However, some keras backend tests failed,  checking now.

Reply:
Some local buffers involved in calculations on the host function, I am working on adding logic to detect these buffers and copy them from/to device.

Example 1:
testSwitch: need to copy buffer %0 from host to device.
```
   func @get_value(%arg0: memref<2x3xf32> {tile.name = "anon"}, %arg1: memref<2x3xf32> {tile.name = "anon_0"}, %arg2: memref<f32> {tile.name = "anon_1"}, %arg3: memref<2x3xf32>) {
   %c1_i32 = constant 1 : i32
   %c0_i32 = constant 0 : i32
    %c1 = constant 1 : index
    %c2 = constant 2 : index
    %c3 = constant 3 : index
    %0 = alloc() : memref<i32>
    %1 = load %arg2[] : memref<f32>
    %2 = fptoui %1 : f32 to i1
    %3 = select %2, %c1_i32, %c0_i32 : i32
    store %3, %0[] : memref<i32>
    "gpu.launch_func"(%c1, %c1, %c1, %c2, %c3, %c1, %0, %arg1, %arg0, %arg3) {kernel = @get_value_kernel::@get_value_kernel} : (index, index, index, index, index, index, memref<i32>, memref<2x3xf32>, memref<2x3xf32>, memref<2x3xf32>) -> ()
    return
  }
```

Example 2:
testAvgPool: need to copy buffer %1 and %2 from device to host.
```
  func @get_value(%arg0: memref<1x4x4x1xf32> {tile.name = "anon"}, %arg1: memref<1x1x1x1xf32>) {
    %cst = constant 0.000000e+00 : f32
    %c0_i32 = constant 0 : i32
    %c0 = constant 0 : index
    %c4 = constant 4 : index
    %c1 = constant 1 : index
    %0 = alloc() : memref<1x4x4x1xi32>
    "gpu.launch_func"(%c1, %c1, %c1, %c4, %c4, %c1, %0) {kernel = @get_value_kernel::@get_value_kernel} : (index, index, index, index, index, index, memref<1x4x4x1xi32>) -> ()
    "gpu.launch_func"(%c1, %c1, %c1, %c4, %c4, %c1, %0) {kernel = @get_value_kernel_0::@get_value_kernel} : (index, index, index, index, index, index, memref<1x4x4x1xi32>) -> ()
    %1 = alloc() : memref<1x1x1x1xi32>
    store %c0_i32, %1[%c0, %c0, %c0, %c0] : memref<1x1x1x1xi32>
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c1, %c1, %0, %1) {kernel = @get_value_kernel_1::@get_value_kernel} : (index, index, index, index, index, index, memref<1x4x4x1xi32>, memref<1x1x1x1xi32>) -> ()
    %2 = alloc() : memref<1x1x1x1xf32>
    store %cst, %2[%c0, %c0, %c0, %c0] : memref<1x1x1x1xf32>
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c1, %c1, %arg0, %2) {kernel = @get_value_kernel_2::@get_value_kernel} : (index, index, index, index, index, index, memref<1x4x4x1xf32>, memref<1x1x1x1xf32>) -> ()
    %3 = load %2[%c0, %c0, %c0, %c0] : memref<1x1x1x1xf32>
    %4 = load %1[%c0, %c0, %c0, %c0] : memref<1x1x1x1xi32>
    %5 = sitofp %4 : i32 to f32
    %6 = divf %3, %5 : f32
    store %6, %arg1[%c0, %c0, %c0, %c0] : memref<1x1x1x1xf32>
    return
  }
```

Reply:
Thanks @jbruestle, the user buffers and the issues above are resolved.